### ISSUE TITLE
Slow download speed (Using Elementum)

### ISSUE THREAD

**Issue by burekas7** [anacrolix/torrent#234]
Hi,

I'm using Elementum 0.0.39, have tried the two options of: File Storage and Memory Storage.
On both I got very low speed.

* File Storage (Max download limit = 8000kbs, Connections limit = 150, Init buffer = 30MB)
Speed = ~500kbs - 1200kbs
Kodi log: https://paste.ubuntu.com/26496607/
Torrent log: https://paste.ubuntu.com/26496608/

* Memory Storage (Max download limit = 9000kbs, Connections limit = 150, Init buffer = 100MB, Memory size for use = 250Mb)
Speed = ~300kbs - 400kbs
Kodi log: https://paste.ubuntu.com/26496618/
Torrent log: https://paste.ubuntu.com/26496622/

Kodi 17.6
Windows 10, 64 bit
@elgatito

**Comment by hoshsadiq** [comment]
It's worth referring to the other post #55
By the way, I can't help but feel my system is having slow speed issues as well. I have a gigabit/s speed connection, however, some torrents seem to be slow despite having plenty of seeders.

**Comment by anacrolix** [comment]
Thanks for the report, I'm taking a look.

**Comment by elgatito** [comment]
Good to mention that Elementum is build with xgo/1.9.2, so is always using CGO. 
When memory storage runs good and file storage is slow, it's obvious that there is an issue with Android/Files management.

More interesting when there are issues for memory storage as well. 
But it's too hard to find the cause. Maybe cmd/torrent/ can be compiled for android and provided the users to see if it shows same low speed. 

I'm also not sure about the influence of Android NDK version on this (not familiar with all that stuff). xgo uses pretty old one.

**Comment by burekas7** [comment]
I'm using windows.

@elgatito 
Do you have any reference torrent that you sure about it and get good download rates with it? 
So I will do the tests with it for the next times?

**Comment by burekas7** [comment]
There is no improvements, I have tested it now with Elementum 0.0.41, and the status is pretty the same: On both, File Storage and Memory Storage, I got very low speed and unstable.

* File Storage (Max download limit = 8000kbs, Connections limit = 150, Init buffer = 30MB)
Speed = ~500kbs - 1200kbs
Kodi log: https://paste.ubuntu.com/26520368/
Torrent log: https://paste.ubuntu.com/26520359/

* Memory Storage (Max download limit = 9000kbs, Connections limit = 150, Init buffer = 100MB, Memory size for use = 250Mb)
Speed = ~700kbs - 1500kbs (For a few second it was ~2000kbs)
Kodi log: https://paste.ubuntu.com/26520462/
Torrent log: https://paste.ubuntu.com/26520454/

***
And by the way, where are all the seeds/peers that shown in the 'Available streams results'?
Why the torrent status is showing very low amount of seeds?
![tor1](https://user-images.githubusercontent.com/13424257/35781210-aeb9abec-09de-11e8-8234-5fccb524c718.png)

**Comment by elgatito** [comment]
Is there any firewall?
Please, try without limits.

**Comment by burekas7** [comment]
The windows defender firewall is active, but it also active when I'm using the Quasar engine and I don't any issues. I will try to remove the download limit and I will update.

**Comment by burekas7** [comment]
@elgatito 
There is no a big difference.

**Without download limit:**

• File Storage (Max download limit = null, Connections limit = 150, Init buffer = 30MB)
 Speed = ~500kbs - 1500kbs
 Kodi log: https://paste.ubuntu.com/26521075/
 Torrent log: https://paste.ubuntu.com/26521068/


• Memory Storage (Max download limit = null, Connections limit = 150, Init buffer = 100MB, Memory size for use = 250Mb)
 Speed = ~400kbs - 1000kbs
 Kodi log: https://paste.ubuntu.com/26521144/
 Torrent log: https://paste.ubuntu.com/26521135/
***
**All windows firewalls are disable (Area, private, public)**
• File Storage (Max download limit = null, Connections limit = 150, Init buffer = 30MB)
 Speed = ~500kbs - 1500kbs
 Kodi log: https://paste.ubuntu.com/26521283/
 Torrent log: https://paste.ubuntu.com/26521276/

• Memory Storage (Max download limit = null, Connections limit = 150, Init buffer = 100MB, Memory size for use = 250Mb)
 Speed = ~500kbs - 1300kbs
 Kodi log: https://paste.ubuntu.com/26521214/
 Torrent log: https://paste.ubuntu.com/26521205/

**Comment by anacrolix** [comment]
I don't see anything clearly wrong in the torrent log. What's your CPU usage look like?

**Comment by anacrolix** [comment]
This one is interesting: go-libutp: 2018/02/04 23:27:26 socket.go:134: ignoring socket read error: read udp [::]:62090: wsarecvfrom: A message sent on a datagram socket was larger than the internal message buffer or some other network limit, or the buffer used to receive a datagram into was smaller than the datagram itself.

What kind of internet connection do you have? Are you on fiber or cable of some kind without a modem?

**Comment by burekas7** [comment]
> I don't see anything clearly wrong in the torrent log. What's your CPU usage look like?
What do you mean? To check the CPU usage of Elementum.exe while it's downloading?

> What kind of internet connection do you have? Are you on fiber or cable of some kind without a modem?
It's cables with modem.
Max. Download speed should be ~100MBs, but I'm using Wifi so it's ~40-50Mbs max.
(Again, I don't know how it's implemented in Quasar engine, but there I get download rates of 4000kbs-5000kbs with good and popular torrents. Maybe also one of the reasons is that they are also using the peers and not only seeds)

**Comment by anacrolix** [comment]
Yes, please check CPU and speeds, when @elgatito makes his next release that incorporates https://github.com/anacrolix/torrent/commit/05da46ddbb02845b1af95cf6529de55cd3c0ce5b and https://github.com/anacrolix/go-libutp/commit/a3dde14e2fdc5c71ccb4c7bbaed7bb58f936a6d0.

**Comment by burekas7** [comment]
OK, I will update you when it will be ready.

By the way, does it still matter if the windows firewall is active, and also if there are down/up limits?

**Comment by elgatito** [comment]
> By the way, does it still matter if the windows firewall is active, and also if there are down/up limits?

It's always better to measure without any firewalls/limits, to identify the problem.

**Comment by elgatito** [comment]
@burekas7  you can try the 0.0.42 with latest changes.

**Comment by burekas7** [comment]
Results tests for 0.0.42
Much better but it's still not so stable and seems that it doesn't work so well when the windows firewall is on.

**All windows firewalls are disable (Area, private, public) + No speed limits**

• File Storage (Max download/upload limit = null, Connections limit = 150, Init buffer = 30MB)
 Speed = ~1000kbs - 3500kbs
 Kodi log: https://paste.ubuntu.com/26525303
 Torrent log: https://paste.ubuntu.com/26525301

• Memory Storage (Max download/upload limit = null, Connections limit = 150, Init buffer = 150MB, Memory size for use = 250Mb)
 Speed = ~2000kbs - 4000kbs
 Kodi log: https://paste.ubuntu.com/26525409
 Torrent log: https://paste.ubuntu.com/26525414
***
**With windows defender firewall, without download/upload limits**

• File Storage (Max download/upload limit = null, Connections limit = 150, Init buffer = 30MB)
 Speed = ~800kbs - 2000kbs  (It took very long time until it start moving and getting speed)
 (Also got a timeout for buffering the initial 30MB : https://paste.ubuntu.com/26525876)
 Kodi log: https://paste.ubuntu.com/26526038
 Torrent log: https://paste.ubuntu.com/26526032

• Memory Storage (Max download/upload limit = null, Connections limit = 150, Init buffer = 150MB, Memory size for use = 250Mb)
 Speed = ~1500kbs - 3500kbs
 Kodi log: https://paste.ubuntu.com/26525713
 Torrent log: https://paste.ubuntu.com/26525705
***
**With windows defender firewall and with download/upload limits**

• File Storage (Max download = 8000kbs, Max upload limit = 500kbs, Connections limit = 150, Init buffer = 30MB)
 Speed = ~800kbs - 1700kbs (It took very long time until it start moving and getting speed)
 Kodi log: https://paste.ubuntu.com/26525965
 Torrent log: https://paste.ubuntu.com/26525960

• Memory Storage (Max download = 8000kbs, Max upload limit = 500kbs, Connections limit = 150, Init buffer = 150MB, Memory size for use = 250Mb)
 Speed = ~1500kbs - 2500kbs
 Kodi log: https://paste.ubuntu.com/26525861
 Torrent log: https://paste.ubuntu.com/26525857

**Comment by burekas7** [comment]
Open questions:
* If there is upload activation so where are the peers amount next to the seeds? (Like Quasar)
 - Seeders I mean by users that have 100% torrent file
 - Peers I mean by users that have part of the torrent file
* Why the "S" (seeders) number are so low related to the number that shown in the available streams results dialog (As I asked here: https://github.com/anacrolix/torrent/issues/234#issuecomment-362931179)

**Comment by elgatito** [comment]
> Why the "S" (seeders) number are so low related to the number that shown in the available streams results dialog (As I asked here: #234 (comment))

Ammount that you get from tracker website is what they write, nobody checks that.

Ammount that you see in torrent stats is the real number we get and verified. It can be both bigger and smaller.

**Comment by elgatito** [comment]
> Results tests for 0.0.42

What about the CPU usage by `Elementum.exe`?

**Comment by burekas7** [comment]
> What about the CPU usage by Elementum.exe?

**Comment by burekas7** [comment]
> What about the CPU usage by Elementum.exe?

**Comment by elgatito** [comment]
The quicker one.

**Comment by burekas7** [comment]
OK, have been checked on this case:
**All windows firewalls are disable (Area, private, public) + No speed limits**
**Memory Storage**
(Max download/upload limit = null, Connections limit = 150, Init buffer = 150MB, Memory size for use = 250Mb)
 Speed = ~1500kbs - 3500kbs

CPU:
20-25% on idle.
14-20% when downloading

Is it possible that on downloading it less then on idle?

**Comment by burekas7** [comment]
OK, have been checked on this case:
**All windows firewalls are disable (Area, private, public) + No speed limits**
**Memory Storage**
(Max download/upload limit = null, Connections limit = 150, Init buffer = 150MB, Memory size for use = 250Mb)
 Speed = ~1500kbs - 3500kbs

**kodi.exe CPU usage:**
20-25% on idle.
14-20% when downloading
* Is it possible that on downloading it less then on idle?

**Comment by burekas7** [comment]
OK, have been checked on this case:
**All windows firewalls are disable (Area, private, public) + No speed limits**
**Memory Storage**
(Max download/upload limit = null, Connections limit = 150, Init buffer = 150MB, Memory size for use = 250Mb)
 Speed = ~1500kbs - 3500kbs

**Total CPU usage:**
10-20% on idle.
10-20% when downloading

**kodi.exe CPU usage:**
20-25% on idle.
14-20% when downloading

**Comment by elgatito** [comment]
@burekas7 can you try to disable UTP in plugin settings and run a quick test?

**Comment by burekas7** [comment]
OK I will check. With the same case right?

**Comment by elgatito** [comment]
> OK I will check. With the same case right?

Yes, try with the fastest case to see if it makes a difference in any way.

**Comment by elgatito** [comment]
@burekas7 , also can you open `http://IP:65220/debug/vars` and paste it's content?

**Comment by burekas7** [comment]
OK, have been checked on this case:
All windows firewalls are disable (Area, private, public) + No speed limits
Memory Storage
 (Max download/upload limit = null, Connections limit = 150, Init buffer = 150MB, Memory size for use = 250Mb)
UTP - Disabled
 Speed = ~3500kbs - 4000kbs (Pretty stable)
Elementum 0.0.42
Kodi 17.6
Windows 10, 64bit

Kodi log: https://paste.ubuntu.com/26530830
Torrent log: https://paste.ubuntu.com/26530832
Torrent debug vars: https://paste.ubuntu.com/26530839

Is it good?
This UTP setting could be improve also the "file storage" option?
***
@elgatito 
By the way, on the first time I tried I had a strange crash of Elementum addon in kodi, you can see it in this kodi kog: https://paste.ubuntu.com/26530763

**Comment by burekas7** [comment]
> Ammount that you get from tracker website is what they write, nobody checks that.
> Ammount that you see in torrent stats is the real number we get and verified. It can be both bigger and smaller.

So if I'm understand it right, those total numbers of seeds/peers in Quasar download status bar and also in the available streams results dialog (Also here in Elementum) are actually fake and isn't checked?

**Comment by elgatito** [comment]
@anacrolix Panic from a dht server:

```
18:27:28.325 T:1100  NOTICE: [plugin.video.elementum] panic: read udp [::]:5813: wsarecvfrom: The connection has been broken due to keep-alive activity detecting a failure while the operation was in progress.
18:27:28.336 T:1100  NOTICE: [plugin.video.elementum]
18:27:28.345 T:1100  NOTICE: [plugin.video.elementum] goroutine 11 [running]:
18:27:28.349 T:1100  NOTICE: [plugin.video.elementum] github.com/anacrolix/dht.NewServer.func1(0xc0421680b8)
18:27:28.355 T:1100  NOTICE: [plugin.video.elementum] 	/ext-go/2/src/github.com/anacrolix/dht/server.go:157 +0x104
18:27:28.361 T:1100  NOTICE: [plugin.video.elementum] created by github.com/anacrolix/dht.NewServer
```

**Comment by elgatito** [comment]
> So if I'm understand it right, those total numbers of seeds/peers in Quasar download status bar and also in the available streams results dialog (Also here in Elementum) are actually fake and not checked or verified?

No. Every torrent client/engine calculates peers number, depending on existing information from different source. You won't see same numbers.

**Comment by burekas7** [comment]
I have checked again my preferred case:

**• File Storage (Max download = 8000kbs, Max upload limit = 500kbs, Connections limit = 150, Init buffer = 30MB)
* With windows defender firewall and with download/upload limits**

And I got this result, it seems that there is improvement:
 Speed = ~800kbs - 4000kbs

But some comments:
* The speed is still not so stable.
* It took very long time until the init buffer is fully loaded, and it's weird because the average speed was around 2000kb/s most of the time and the percentage increase slowly. So I got Timeout error for the buffer init that still haven't reached to 100% buffer.
Also in the beginning it was stuck on a zero (~10-50kbs) for 1,2 minutes and just then it start to gain speed. 

* Kodi log: https://paste.ubuntu.com/26536556
* Torrent log: https://paste.ubuntu.com/26536558
* Torrent debug vars: https://paste.ubuntu.com/26536559

Setup:
Elementum 0.0.42
Kodi 17.6
Windows 10, 64bit (Firewall is enabled)

**Comment by anacrolix** [comment]
I notice you have no connections with RC4 header encryption, and very few with encryption enabled at all. 

    torrent.EncryptionPolicy{DisableEncryption:true, ForceEncryption:false, PreferNoEncryption:false}

That's not a good idea.

**Comment by anacrolix** [comment]
"mseCryptoProvides": {"2": 26, "3": 13},

Two-thirds of peers connected to with message stream encryption are being rejected with encryption disabled. This is probably why disabling uTP gave a speedup, as the rate of encryption on that protocol is much higher.

I also strongly recommend using an IPBlocklist, see https://github.com/anacrolix/confluence/blob/master/main.go#L68.

**Comment by elgatito** [comment]
```
EncryptionPolicy: gotorrent.EncryptionPolicy{
			DisableEncryption: s.config.EncryptionPolicy == 0,
			ForceEncryption:   s.config.EncryptionPolicy == 2,
},
```
Where `config.EncryptionPolicy` can be [Enabled, Disabled, Forced]. It's controlled by user in Kodi settings. Default is "enabled".

https://github.com/elgatito/elementum/blob/master/bittorrent/service.go#L178 - blocklist is used.

**Comment by elgatito** [comment]
@burekas7 , try to enable encription in Bittorrent tab of plugin settings.

**Comment by burekas7** [comment]
And keep the UTP disabled?

**Comment by elgatito** [comment]
@burekas7 you can try both to see the difference.

**Comment by burekas7** [comment]
@elgatito 
I have checked the settings and the encryption was enabled. Are you sure it wasn't by the log?
I will do the test again.

**Comment by burekas7** [comment]
The same, the speed is not the best and also not stable.

• **All windows firewalls are disable** (Area, private, public) + No speed limits
• **Memory Storage**
 (Max download/upload limit = null, Connections limit = 150, Init buffer = 150MB, Memory size for use = 250Mb)
• **UTP** - Disabled
• **DHT** - Enabled
• **Encryption policy** - Enabled
• **Download Speed** = ~1000kbs - 2500kbs
• **Logs:**
 * Kodi log: https://paste.ubuntu.com/26542604
 * Torrent log: https://paste.ubuntu.com/26542606
 * Torrent debug vars: https://paste.ubuntu.com/26542612
***
• **All windows firewalls are disable** (Area, private, public) + No speed limits
• **Memory Storage**
 (Max download/upload limit = null, Connections limit = 150, Init buffer = 150MB, Memory size for use = 250Mb)
• **UTP** - Enabled
• **DHT** - Enabled
• **Encryption policy** - Enabled
• **Download Speed** = ~500kbs - 1700kbs
• **Logs:**
 * Kodi log: https://paste.ubuntu.com/26542546
 * Torrent log: https://paste.ubuntu.com/26542538
 * Torrent debug vars: https://paste.ubuntu.com/26542554
***
• **All windows firewalls are disable** (Area, private, public) + No speed limits
• **File Storage**
 (Max download = 8000kbs, Max upload limit = 500kbs, Connections limit = 150, Init buffer = 30MB)
• **UTP** - Disabled
• **DHT** - Enabled
• **Encryption policy** - Enabled
• **Download Speed** = ~500kbs - 2000kbs
• **Logs:**
 * Kodi log: https://paste.ubuntu.com/26542704
 * Torrent log: https://paste.ubuntu.com/26542714
 * Torrent debug vars: https://paste.ubuntu.com/26542717
***
Setup:
• Elementum 0.0.42
• Kodi 17.6
• Windows 10, 64bit (Firewall is enabled)

***
By the way, why with Memory Storage the download begin to download and get speed imediatlly, and with File Storage it takes ~1-2 minutes?
Torrent log for File Storage during these ~1-2 minutes: https://paste.ubuntu.com/26542732

**Comment by elgatito** [comment]
> By the way, why with Memory Storage the download begin to download and get speed imediatlly, and with File Storage it takes ~1-2 minutes?

Because with memory storage you don't download all the file at once, you set to download only 1/3 of memory storage, that is why you don't get maximum rate for memory, but you can see the peak rate, which usually is about the full bandwidth.

**Comment by elgatito** [comment]
Strange, log says encryption is disabled.

```
EncryptionPolicy:torrent.EncryptionPolicy{DisableEncryption:true, ForceEncryption:false, PreferNoEncryption:false}
```

**Comment by burekas7** [comment]
So there is nothing else that can be done?
Did you see any change in the last logs?

**Comment by elgatito** [comment]
> So there is nothing else that can be done?

Set Encryption to "Disabled" (not Force), it will become enabled. I will fix it in the next version. 
But you can test with that option.

**Comment by burekas7** [comment]
Much better.
Even when the window firewalls are enabled.
UTP enabled has better results.
Encryption policy disabled (In the addon) gives better results.

• **All windows firewalls are Enabled** (Area, private, public) + No speed limits
• **Memory Storage**
 (Max download/upload limit = null, Connections limit = 150, Init buffer = 250MB, Memory size for use = 250Mb)
• **UTP** - Disabled
• **TCP** - Enabled
• **DHT** - Enabled
• **Encryption policy** - Disabled (In the addon settings)
• **Download Speed** = ~3000kbs - 4000kbs
• **Logs:**
 * Kodi log: https://paste.ubuntu.com/26545471
 * Torrent log: https://paste.ubuntu.com/26545479
 * Torrent debug vars: https://paste.ubuntu.com/26545482
***
• **All windows firewalls are Enabled** (Area, private, public) + No speed limits
• **Memory Storage**
 (Max download/upload limit = null, Connections limit = 150, Init buffer = 250MB, Memory size for use = 250Mb)
• **UTP** - Enabled
• **TCP** - Enabled
• **DHT** - Enabled
• **Encryption policy** - Disabled (In the addon settings)
• **Download Speed** = ~3500kbs - 4200kbs
• **Logs:**
 * Kodi log: https://paste.ubuntu.com/26545502
 * Torrent log: https://paste.ubuntu.com/26545509
 * Torrent debug vars: https://paste.ubuntu.com/26545514
***
• **All windows firewalls are Enabled** (Area, private, public) + With speed limits
• **File Storage**
 (Max download = 8000kbs, Max upload limit = 500kbs, Connections limit = 150, Init buffer = 30MB)
• **UTP** - Enabled
• **TCP** - Enabled
• **DHT** - Enabled
• **Encryption policy** - Disabled (In the addon settings)
• **Download Speed** = ~1000kbs - 3500kbs (Less stable then Memory Storage, but most of the time ~2000kb/s)
• **Logs:**
 * Kodi log: https://paste.ubuntu.com/26545547
 * Torrent log: https://paste.ubuntu.com/26545548
 * Torrent debug vars: https://paste.ubuntu.com/26545551
***
Setup:
• Elementum 0.0.42
• Kodi 17.6
• Windows 10, 64bit (Firewall is enabled)

**Comment by burekas7** [comment]
> Because with memory storage you don't download all the file at once, you set to download only 1/3 of memory storage, that is why you don't get maximum rate for memory, but you can see the peak rate, which usually is about the full bandwidth.
OK, but sometimes because of that, the Timeout error of the init buffer to get loaded fully could happen especially when it's a big torrent file. If I understand what you said.

Is there any way to make the buffer download to start more quickly? (When using File Storage)

**Comment by burekas7** [comment]
@elgatito 
> Set Encryption to "Disabled" (not Force), it will become enabled. I will fix it in the next version.

In Elementum 0.0.43 how should I set this now? Enable/Disable/Force?

**Comment by elgatito** [comment]
Nope, set as it should be. Enabled for enabling.

**Comment by burekas7** [comment]
Encryption policy setting is now ok when enabled.
But the speed is not so best again, even when disable the firewall and without speed limits (I have to mention that the torrent here is not the same from previous tests)

Do you see any change or new helpful info from these logs?
***
• ** Windows firewalls are disabled** (Area, private, public) + No speed limits
• **Memory Storage**
 (Max download/upload limit = null, Connections limit = 150, Init buffer = 250MB, Memory size for use = 250Mb)
• **UTP** - Enabled
• **TCP** - Enabled
• **DHT** - Enabled
• **Encryption policy** - Enabled (In the addon settings)
• **Download Speed** = ~1000kbs - 2400kbs
• **Logs:**
 * Kodi log: https://paste.ubuntu.com/p/FD33tGgtG5
 * Torrent log: https://paste.ubuntu.com/p/fmMbrVsxVK
 * Torrent debug vars: https://paste.ubuntu.com/p/X7tRty6FGt
***
• ** Windows firewalls are disabled** (Area, private, public) + No speed limits
• **Memory Storage**
 (Max download/upload limit = null, Connections limit = 150, Init buffer = 250MB, Memory size for use = 250Mb)
• **UTP** - Enabled
• **TCP** - Enabled
• **DHT** - Enabled
• **Encryption policy** - Disabled (In the addon settings)
• **Download Speed** = <500kbs
• **Logs:**
 * Kodi log: https://paste.ubuntu.com/p/c2Tw6BR5WK
 * Torrent log: https://paste.ubuntu.com/p/CVhP2WrcsB
 * Torrent debug vars: https://paste.ubuntu.com/p/y8V8CfgJDC
***
• **All windows firewalls are Enabled** (Area, private, public) + With speed limits
• **File Storage**
 (Max download = 8000kbs, Max upload limit = 500kbs, Connections limit = 150, Init buffer = 30MB)
• **UTP** - Enabled
• **TCP** - Enabled
• **DHT** - Enabled
• **Encryption policy** - Enabled (In the addon settings)
• **Download Speed** = ~800kbs - 2000kbs (Less stable then Memory Storage, but most of the time ~2000kb/s)
• **Logs:**
 * Kodi log: https://paste.ubuntu.com/p/VCwDqYqHbX
 * Torrent log: https://paste.ubuntu.com/p/nVTvKFyH3z
 * Torrent debug vars: https://paste.ubuntu.com/p/Kk9j4yDzSM
***
Setup:
• Elementum 0.0.43
• Kodi 17.6
• Windows 10, 64bit

**Comment by burekas7** [comment]
@elgatito 
I saw that you wrote that you succeeding to get 100mbs .
So what is the problem with my setup that I don't get better download results?
Does it because the OS (Windows)? Firewall (which disabled)? The router settings? The internet provider? Elementum/Kodi settings? The seeds? etc...

Can you give me kind of a 'golden' torrent so I would check it?

**Comment by elgatito** [comment]
@burekas7 I can get full speed even with 10-20 peers, you don't need 100-150 to use all the bandwidth. I don't know what is the problem with your case.

**Comment by burekas7** [comment]
The logs doesn't show something?

**Comment by burekas7** [comment]
@anacrolix 
Thanks man! It helps :)
I reduced the number of the connections limit from 150 to 60, and now the speed pretty stable and better: ~3000 kbs to 5000 kb/s (4000 kbs is the average)

But, for some reason, now the "File Storage" type gives better result then "Memory Storage", even the Firewalls and the Speed limits were activated when testing the "File Storage", and in "Memory Storage" they were disabled.

By the way, how can I know the optimum limit for my connection? Each time increase it by ten and see if the speed slower?

***
• **Windows firewalls - Enabled** (Area, private, public)
• **With speed limits**
• **File Storage**
 (Max download = 8000kbs, Max upload limit = 500kbs, Connections limit = 60, Init buffer = 30MB)
• **UTP** - Enabled
• **TCP** - Enabled
• **DHT** - Enabled
• **Encryption policy** - Enabled (In the addon settings)
• **Download Speed** = ~3000kbs - 5000kbs
• Logs:
 • Kodi log: https://paste.ubuntu.com/p/ZqmJmMVxr5
 • Torrent log: https://paste.ubuntu.com/p/7RSydTmmN8
 • Torrent debug vars: https://paste.ubuntu.com/p/VJ8w7znkC2
***
• **Windows firewalls - Disabled** (Area, private, public)
• **Without speed limits**
• **Memory Storage**
 (Max download/upload limit = null, Connections limit = 60, Init buffer = 250MB, Memory size for use = 250Mb)
• **UTP** - Enabled
• **TCP** - Enabled
• **DHT** - Enabled
• **Encryption policy** - Enabled (In the addon settings)
• **Download Speed** = ~2000kbs - 3000kbs
• **Logs:**
 * Kodi log: https://paste.ubuntu.com/p/GTCYbJXh2p
 * Torrent log: https://paste.ubuntu.com/p/fPr7vrKm96
 * Torrent debug vars: https://paste.ubuntu.com/p/vwHGvSj9zg
***
Elementum 0.0.43
Kodi 17.6
Win 10, 64bit

**Comment by anacrolix** [comment]
Again, just based on my years of torrenting (:P), I wouldn't set upload to more than half your upload capacity. Increasing upload isn't going to improve your download speeds though.  Try http://www.speedtest.net/ to determine internet performance, can you give that a go and report the results?

The memory storage isn't provided in this repo, while the file storage is. I can't speak to its implementation, but limiting your available readahead, and aggressively dropping pieces as is required with a smaller cache is always going to harm performance.

If you're still keen to try configurations you might try:

1) Setting upload to half your upload capacity. (see the results from speedtest.net)

2) Disable TCP and upload limit, raise connections to ~200.

But your results already look much better, I'm happy to close at this point once we have your speedtest.net results.

I also note from your logs that the half-open limit is absurdly high, I think I'll reduce its default. In addition to your 60 active connections, you have another 80 (potentially 160 if you consider the simultaneously TCP/uTP dial).

**Comment by xdeee** [comment]
Same issue for me. Ubuntu 17.10, Kodi 17.6

**Comment by anacrolix** [comment]
@burekas7 Your speedtest results would help immensely here.

**Comment by burekas7** [comment]
Wifi connection:
![speed_test](https://user-images.githubusercontent.com/13424257/36492173-6d12b5c8-1724-11e8-9e74-8e5ed4a9eaa8.png)

**Comment by burekas7** [comment]
By the way, the upload in Elementum is very very low <10kb/s.
Does it seeding at all during the download with other peers?

Does it have separation now between seeds and peers in the torrent engine?
(Seeds - Users that have 100% torrent file, Peers - User that have part of the torrent file)

**Comment by anacrolix** [comment]
Your upload bandwidth is incredibly bad. I would switch off uploads completely in your settings, you can't afford to spend any of that upload on anything but protocol overhead, especially when you are downloading at very high rates. Yes I believe @elgatito exposed those things in the UI recently from https://godoc.org/github.com/anacrolix/torrent#TorrentStats.

**Comment by ImAGitHubber** [comment]
Hello,

First of all, a big thank you to both elgatito and anacrolix for providing these plugins to the community and helping people with their problems.

Unfortunately, I'm experiencing a very similar issue and it's very frustrating because I cannot get Elementum to work properly on my Android Kodi Box using my USB stick as the file/torrent storage medium.

It was working fine with the old Quasar plugin with higher and more stable speeds.

-----------------------
Android Box:  Minix U1 (http://minix.com.hk/en/products/neo-u1)

USB Stick:  Samsung MUF-128BB/AM (https://www.samsung.com/us/computing/memory-storage/usb-flash-drives/usb-3-0-flash-drive-fit-128gb-muf-128bb-muf-128bb-am/)

Internet:  300mb Fiber Optic connection via Gigabet Ethernet and with the IP DMZ'd.
-----------------------

I have a high-quality android box, USB stick, and internet connection, so I do not believe any of them are the culprit here.

I spent over 4 hours yesterday trying different settings in Elementum but could not get any to work properly.  The speed always jumps all over the place and the files do not play smoothly, especially if I jump backwards (or forwards) when the video plays.  The highest I've seen the speed go is about 5000 kb/s, but the average is closer to around 400 kb/s.

So I set everything to the "Default" settings in Elementum and set the file storage to be a folder in my USB stick and ran the following test:

-----------------------
Here is the Kodi Log file:  http://ix.io/10tS

Here is video I took while making that log so you can see how Elementum behaves very erratically with the torrenting:    https://streamable.com/przse
-----------------------

As you can see in the video above, something is terribly wrong.

Please help me fix this so it runs smoothly.  Thank you.

**Comment by anacrolix** [comment]
@ImAGitHubber Thanks for the very detailed info, I'm taking a look.

**Comment by anacrolix** [comment]
@elgatito can we get the /debug/perf output? I want to see the writeChunk performance, here's how it looks on my machines:
```
description  count    min       mean       max
write chunk  2471948  98.679µs  489.624µs  432.511787ms
```

**Comment by ImAGitHubber** [comment]
@anacrolix You are welcome.  Please let me know if you or @elgatito need anything else from my end to debug and fix this.  Thanks again.

**Comment by elgatito** [comment]
> @elgatito can we get the /debug/perf output?

@anacrolix what is that? how to get it?

**Comment by anacrolix** [comment]
The perf stuff is exposed through [/debug/perf on the default http handler](https://github.com/anacrolix/missinggo/blob/master/perf/events.go#L18).

**Comment by jhill54** [comment]
@ImAGitHubber @anacrolix Were you guys able to find a solution?  I am also experiencing the same issues of very slow speed vs. Quasar - running on a 200mbit line and new Nvidia Shield.  Thanks for any help!

**Comment by elgatito** [comment]
@jhill54 If you switch to memory storage, set, say, at least 150Mb and see if speed comes to better numbers. Then it's all about default storage engine on some OSs.

**Comment by jhill54** [comment]
Unfortunately it won’t let me change the default storage size on the Shield. I tried some other things and still no luck.

**Comment by ImAGitHubber** [comment]
> @jhill54 If you switch to memory storage, set, say, at least 150Mb and see if speed comes to better numbers. Then it's all about default storage engine on some OSs.

This is a potential workaround to this problem, but not an actual solution.  Plus, the "Memory Storage" option does not save Multi-file Torrents, so you have to keep searching for the torrent every time instead of being able to save it in a proper storage drive and then just opening the next file when you need it.

When can we expect a solution to this issue?

**Comment by jhill54** [comment]
@elgatito thanks for advice, but the download speeds are still extremely slow, and the connection dies after about 5 mins consistently.  The speed drops to 0 kbps, and does not re-initialize.

**Comment by ImAGitHubber** [comment]
> @jhill54 without providing any log files, there is no way to find a problem.

I provided log files and even video of the slow speeds a month and a half ago and I'm still waiting for the problem to be found.

**Comment by elgatito** [comment]
@ImAGitHubber when you have big speeds with memory storage and slow with file storage - it's one issue, and it's described above. When you have bad speeds with any storage (this can be checked only with log files), like @jhill54 - it's another issue.

**Comment by ImAGitHubber** [comment]
> @ImAGitHubber when you have big speeds with memory storage and slow with file storage - it's one issue, and it's described above.

Thank you for your reply.  Where is my problem described?

@anacrolix had requested from you the /debug/perf output to help debug my issue, but it doesn't you ever provided it.

Please advise.  Thanks again.

**Comment by elgatito** [comment]
@ImAGitHubber  @jhill54 Latest Elementum 0.0.47 includes `/debug/` url, so please, post here your debug information, so @anacrolix could try to find the problem.

**Comment by elgatito** [comment]
There is a new `/debug/bundle` url in elementum 0.0.48, which downloads all the debug information from elementum/kodi. 
Can anybody test on android and post results here?

http://192.168.0.1:65220/debug/bundle (replace 192.168.0.1 with your Kodi IP)

**Comment by burekas7** [comment]
@elgatito 
I still get error with 0.0.49 version as described already in some issues with the last versions (File Storage). But I will try to check it when there will be a working version.

And by the way, to be honest, I had back to Quasar which is much better and stable for a long time and never disappoints (no offend) .
Also the Trakt support is still not totally good (But this is for another issue)

**Comment by elgatito** [comment]
@burekas7 I was asking for a log file and debug information. 

It's good you have found plugin that suites you :+1:

**Comment by xdeee** [comment]
Suddenly it started to download at full speed for me about a month ago. I did nothing. Even didn't update plugin or OS. I'm not sure, but probably I reset Elementum's settings.
Now it works very well both on Ubuntu and on Android box. Thank you.

**Comment by elgatito** [comment]
@xdeee by default, Elementum now is using Memory storage, so you have better speeds. But if you switch to File storage and has problems - it would be useful to get a `/debug/bundle` log file from you.

**Comment by xdeee** [comment]
I had speed issues no matter which type of storage I used. I'll try switching to File though and report debug info in case of any troubles

**Comment by anacrolix** [comment]
Might be related to #216. If we can get evidence that slow write speeds are the issue here, I'll look into create a write cache like libtorrent does (or how I assume it does). This will be easier now that the [request strategy is more selective](https://github.com/anacrolix/torrent/issues/253#issuecomment-400887420), the writes can be done linearly and should help immensely with caching.

**Comment by anacrolix** [comment]
Thanks @XavierCLL . It's clear from your log, that something is not right with the lock times. Locks are being held for absurdly long periods, this will severely limit network throughput. There's something different about that system, or the configuration in use.

**Comment by anacrolix** [comment]
Here are mine for comparison. The averages are all well below 1ms, with maximums normally several 100ms (usually due to GC or contention I believe). Your averages are 10s of ms, with peaks up to 17s. As a simple example, that would limit receiving bandwidth to much less than 1MB/s.

**Comment by btrrj** [comment]
Hi, i saw this topic and i'm having also troubles with Elementum download Speed.
In Quasar was ok, almost ALL the time it was using all my bandwidth ir near it.
I uninstalled Quasar thinking i would have better performance with Elementum...

I used almost every settings o used i Quasar. I'm using an external USB 3.0 HD to cache files.

Using Elementum 0.0.53 in my Android set top box.

Any tip?

**Comment by anacrolix** [comment]
Try the memory storage. Is the USB HD an SSD or HDD?

**Comment by btrrj** [comment]
@anacrolix It´s a 500gb Momentus XT SSHD, it´s quite fast...
But like i said, in Quasar it always worked perfectly! I uninstalled Elementum and reinstalled Quasar, tested the same movie using the same source, voilá: 5mb/s (in Elementum i got 60kb/s MAXIMUM)

It´s something wrong with the app for sure.

**Comment by XavierCLL** [comment]
I have the same situation as @btrrj, in Quasar work fine at maximum speed, in Elementum ~5-10% of the network, testing the same torrent

**Comment by joaopauloss** [comment]
Same problema when using elementum.  Created an issue  on elementum's git but  @elgatito closed and referenced this ond. 
So,  quasar speed is ok but elementum  is about 8% top speed. If need new logs,  please  tell me.

https://github.com/elgatito/plugin.video.elementum/issues/238#issuecomment-413242003

**Comment by btrrj** [comment]
I just gave up Elementum for now and got Quasar back...Sorry, but It works much better. It's a shame Quasar is not updated anymore.
Speed is better in 0.61, but not in par with Quasar yet. Also, everytime i get  "no links were found" in recent movies. 
I will wait further Elementum development and teste It again later.

**Comment by joaopauloss** [comment]
Yeah.  Elementum was the best for me because my Android TV box doesn't have too much rom and never get good speed with quasar when using a external flash drive.  So elementum using "memory" storage type was perfect.  Unfortunately, it worked just a few times.

**Comment by elgatito** [comment]
@anacrolix  there are few feedbacks from users. Some say they have to turn *OFF* uTP, because their routers are dying, some say their Windows-running host suffers a lot. That is all about memory storage.

The funny thing, earlier releases were not causing a lot of problems (when they were working slower and making less network IO operations). 

 It is probably good for server side to have good parallelism (from Go), but no so good for limited OS and weak routers :)

Maybe we can add an option of doing kind of sleep before creating new connection, not sure how can we fight with that.

**Comment by anacrolix** [comment]
The client does establish connections quite aggressively. Lowering the number of simultaneous torrents loaded, and the ClientConfig.HalfOpenConnsPerTorrent value should directly affect that. https://github.com/anacrolix/torrent/commit/2940f27f9f8d6bd085bfa8a3f75d83e1bb423817 should have helped with overhead caused by incoming connections regarding utp.

**Comment by elgatito** [comment]
@anacrolix Our defaults are now:
```
	EstablishedConnsPerTorrent = 50
	TorrentPeersHighWater = 3000
	HalfOpenConnsPerTorrent = 50
```

I can make "modes" which would bundle those settings into mini/normal/maxi. I would much appreciate if you can give your suggestions about probable numbers for those settings.

I was talking about speed of creating connections. Creating 50 connections per second is one thing, same 50 connections created per 100 msec is another thing...

**Comment by anacrolix** [comment]
@elgatito I think I follow. Perhaps we need a "connection interval" instead, or in addition to, the half-open limit? That would solve it neatly.

**Comment by anacrolix** [comment]
It's hard to say, since we don't know for certain if this is even the reason those users are having difficulties. But we do know it's probably related to UDP on Windows.

**Comment by joaopauloss** [comment]
@anacrolix, i'm getting those issues but don't use windows. Just android: one tv box and smartphone. =/

> It's hard to say, since we don't know for certain if this is even the reason those users are having difficulties. But we do know it's probably related to UDP on Windows.

**Comment by anacrolix** [comment]
@elgatito after some investigation, reducing the peers high water will reduce the impact dht announces have, as announces will terminate earlier when reaching that number of peers. This is just a work around until https://github.com/anacrolix/dht/issues/24 or something like it is implemented.

**Comment by acsterf** [comment]
> @elgatito after some investigation, reducing the peers high water will reduce the impact dht announces have, as announces will terminate earlier when reaching that number of peers. This is just a work around until [anacrolix/dht#24](https://github.com/anacrolix/dht/issues/24) or something like it is implemented.

@elgatito how can we change these settings in Elementum?

**Comment by elgatito** [comment]
@acsterf @anacrolix 

Currently we do set there values to ClientConfig:
```
	s.ClientConfig.EstablishedConnsPerTorrent = s.config.ConnectionsLimit
	s.ClientConfig.TorrentPeersHighWater = 3000
	s.ClientConfig.HalfOpenConnsPerTorrent = 50
```

Where `ConnectionsLimit` is configurable.
Should we also make other values configurable? 

I'm also having issues on my new home network, where router is getting crazy after torrent start, speed comes to 0 kb/s for connections are lost (for all the devices behind that router). Switching to a VPN helped a little bit, but still not the best speed so far.

**Comment by xdeee** [comment]
Turning DHT off also helped a bit making Elementum usable (with 2-3 stops while watching shows). But some torrents don't even start downloading without DHT due to provider-blocked trackers.

**Comment by anacrolix** [comment]
Please try with https://github.com/anacrolix/dht/commit/efcc29278754bdd5e766edfb6675f3fe0199fc6f.

**Comment by elgatito** [comment]
@anacrolix any suggestions for setting up TorrentPeersHighWater or HalfOpenConnsPerTorrent?

**Comment by anacrolix** [comment]
@elgatito I'd put HalfOpenConnsPerTorrent to maybe half of EstablishedConnsPerTorrent, which I'd put at approximately the number of connections you expect the users network to support (100-200 are conservative guesses), divided by the number of torrents you expect them to have loaded simultaneously. #273 and a global connection limit would help this but I haven't done either.

TorrentPeersHighWater mainly affects memory use, and when to stop searching for peers. 1-10x EstablishedConnsPerTorrent would be a reasonable value.

**Comment by Mahava** [comment]
Hi.
What's the Plugin connectivity in the 'Advanced' tab?
Also, is there a way to pass the selected link onto real-debrid to play?  That is, to use RD as the torrent streamer?

**Comment by elgatito** [comment]
@Mahava 'No' to all. And it's not the right place to ask.

**Comment by elgatito** [comment]
I'd be much thankful if anyone could try the latest Elementum 0.0.62, which includes all the latest fixes/changes of the torrent library.

**Comment by zaro** [comment]
I have pine64 with android, and recently reinstalled kodi and switched from Quasar to Elementum (0.0.62) .
Download speed went from the limit of my ISP (~50-Mbps) down to 10-20Mbps (if there is local peer in the same city) and 6-7Mpbs if there is no local peer.
This is on Ethernet, so wifi is definitely not an issue.

**Comment by elgatito** [comment]
@zaro you compare speeds for file storage? or memory storage?

**Comment by zaro** [comment]
File storage.  What is the difference ?

**Comment by elgatito** [comment]
This issue was initially describing issues with download speed even with memory storage.
Which we still, hope, to be retested with latest version of Elementum, by people, who had this issue.

**Comment by xdeee** [comment]
Speed is much better now - about 7Mb/s (of 11Mb/s full bandwidth) and my router feels good when I watch a movie. But video playback became irritating. FPS drops and freezes when downloading is active. CPU load is about 60%.
I'm using Mi Box 3, elementum 63 with memory storage.
bundle: https://drive.google.com/file/d/16qkEwDd70TeGpvebwYjEWTVsk5R1nP7l/view?usp=sharing

**Comment by anacrolix** [comment]
@xdeee Please try again with d9fe3f50d3f17e13bf037fcf0a212bdb769e13dd or later.

**Comment by mbLc** [comment]
Hi,

I got the same issue, my speed is blocked at around 3Mo/s with memory download and with FAT32 it's extremely slow (around 500ko/s).
I'm downloading the same file at 10 / 15Mo/s with the VPN on my computer.

Before the speed was blocked at 2,5Mo/s and since I put the connections limit at 60 and the memory size at maximum it improved the speed.

Here is my debug: 

http://pasted.co/94b1aa3b

**Comment by antonsoroko** [comment]
@anacrolix @elgatito 
Hi!
Unfortunately I have the same issue.
env: elementum 0.66, android 8.0, xiaomi mi box 3, ram as storage

1) log+perf https://paste.ubuntu.com/p/7VqTpSsyMT/
perf https://paste.ubuntu.com/p/PDNqPvQ9d2/
perf https://paste.ubuntu.com/p/SmrtvFxgkS/

2) log https://paste.ubuntu.com/p/zmrwkq8YZS/
perf https://paste.ubuntu.com/p/DsCsDRCjvM/
perf https://paste.ubuntu.com/p/fQRwykWRTV/

Thank you!

**Comment by elgatito** [comment]
@antonsoroko you have 50mb initial buffer, which is too low if you expect to see 100mb speeds, especially if you have no low number of peers. You can set initial buffer to 150mb and enable DHT and set everything to "auto" (connection limit).

**Comment by antonsoroko** [comment]
@elgatito 
It is always the same
I have tried "250MB as memory for torrent + 250MB + DHT + auto" for buffer for many times
maybe some of these logs have different parameters, it is because I am already lost in my trials

**Comment by antonsoroko** [comment]
example for "250MB as memory for torrent + 250MB for buffer + DHT + auto limits" and foreign torrent file:
log+perf https://paste.ubuntu.com/p/5tD3zHvqRk/
perf https://paste.ubuntu.com/p/Nv8FF9w9Td/
perf https://paste.ubuntu.com/p/Rg2HXJKzdf/
perf https://paste.ubuntu.com/p/ZDVpPkX7Ks/

thanks.

**Comment by antonsoroko** [comment]
BTW, I launched elementum daemon on my linux pc (with write to disk) - after it downloads about 500MB  
 downloading speed becomes quite good (in comparison to android).

**Comment by heartlogic** [comment]
Hello...

I am also have low download speed with elementum...

I have installed it in Tanix TX5 Pro box with CoreELEC 8.95.6 (LE 9.0) in SDCard...

I compare it with Quasar installed in LibreELEC with Krypton 17.6 and Quasar starts immediately downloading at the top speed of VDSL line (5000kb/s D - 500kb/s U)...

With Elementum at best I reach 900kb/s Download speed...

I have tested be all options and nothing fixed this issue..

Can you help?

Best regards...

**Comment by acsterf** [comment]
Great work guys, since 0.66 I can achieve my internet throughput speed (abouta30 Mb/s) and memory storage is usable for large torrents.

Regarding how it works, a 200 MB buffer + 200 MB memory storage means:

a) 400 MB of RAM;
B) 200 MB of disk storage + 200 MB of RAM;

**Comment by elgatito** [comment]
I would say this issue needs to be closed.

Initial problem was related to network throughput, not linked with storage speeds. And that seems to be fixed.

File storage speed is tracked here: #265

**Comment by elgatito** [comment]
@heartlogic you should read this issue first. Then, if you have any problems - post log files.

**Comment by antonsoroko** [comment]
>that seems to be fixed
I would say that this is arguable...
I have very poor performance with "geographically far" peers and quite good (not perfect, though) with "geographically near" peers. (Logs are available in my previous message)
No offense.
Of course your result may vary :-)

@heartlogic try to use memory as a storage. File storage has issues, see #265 as @elgatito mentioned above.

**Comment by xdeee** [comment]
> I would say this issue needs to be closed.

Tested on three systems today. All devices use memory storage.

At work, Windows 7 x64 fresh Kodi 17 install, ethernet: 
Elementum utilized 4,2Mb/s out of 5.6Mb/s, the speed seems to be sufficient for streaming but remains unstable. Quasar on the same system shows maximum available speed. Lost logs from this run, sorry.

Mac OS, fresh Kodi 17 install, Wifi 5G:
Elementum starts torrent at 5-6Mb/s for 3-4 seconds, then wifi becomes almost unusable on all devices and speed drops to 700Kb-1.8Mb/s.
debug bundle - https://paste.ubuntu.com/p/ZCyjWvRfhP/ 

Android Xiaomi TV Box, Kodi 17, Wifi 5G:
Sometimes it works quite good. Sometimes when I pause playback I see acceptable speed - about 4-5Mb/s (out of 10Mb/s by Quasar) but  after minute or two playback starts slutter and stops.I have to pause it to fill the buffer. When paused I see about 5Mb/s - this is enough to play 1.7Gb file continuously but after a few minutes fps drops and I have to pause playback again. 
Just tested on this system but no troubles this time. I'll drop logs as soon as I get into trouble.

**Comment by mbLc** [comment]
I'm using memory storage, have download issue and posted my log and no one cared to reply...

**Comment by heartlogic** [comment]
I have used both file and memory storage with same results...

I would say that issue is on my setup but by comparing with other programs i conclude that issue is with Elementum...

Please do not get this the wrong way because I know how hard and time consuming is to program and especially for free but Elementum has issues and download speed is one major issue since we are talking for a torrent program...

**Comment by zynexiz** [comment]
It's definitely a issue with Elementum, and has been from the start. I use Quasar instead because this plugin is pretty useless (the UI itself is really good though). I get at most 1,5-2mb/sec download speed (most times just 50-500kb/sec, and halt half of the times) and with Quasar 10-12mb/sec.

I'm starting to wonder if the traffic is routed thru some kind of proxy we don't know about!?

**Comment by heartlogic** [comment]
Yes, I have slow speeds with memory storage as well... The torrent engine is in issue here...

I can help if you point me the way...

**Comment by elgatito** [comment]
@heartlogic http://elementum.surge.sh/contribute/log/

**Comment by zynexiz** [comment]
@antonsoroko 
The problem with slow/halted downloads is occurring with all the storage options, never gets over about 2,5 mb/sec (most of the time lot less than 1 mb/sec), while Quasar download in >8 mb/sec most of the times (depending on the seeders of course).

Just a thought, why not use Quasars engine that's actually working, until the issue is resolved anyway? As for now, there's no way we can use Elementum because of the slow/halted downloads. It's a petty because the UI itself is lot better than Quasars IMO.

**Comment by xdeee** [comment]
This is definitely the torrent library issue. Tested it with [this client](https://github.com/deranjer/goTorrent) using the same library and got the same result: speed keeps at 0.5-0.7Mb/s with some peaks at 2.5Mb/s. BTW: wifi also dies when I start downloading.

**Comment by antonsoroko** [comment]
@xdeee
Thanks for the sharing.
Could you please try with disabled dht and then with disabled utp?

**Comment by heartlogic** [comment]
I have tried all these options in any combination possible... The result is the same...

**Comment by antonsoroko** [comment]
One of workarounds could be https://github.com/anacrolix/torrent/issues/150#issuecomment-450336726
At least this could mitigate some situations.

**Comment by anacrolix** [comment]
@antonsoroko it won't provide much or any improvement. The current piece strategy will fetch any piece reads are blocked on with the highest urgency. If the client is struggling to buffer ahead of a read it's essentially sequential already.

**Comment by elgatito** [comment]
@xdeee 

> This is definitely the torrent library issue. Tested it with this client using the same library and got the same result: speed keeps at 0.5-0.7Mb/s with some peaks at 2.5Mb/s. BTW: wifi also dies when I start downloading.

If your wifi dies - then your networking hardware is crappy and can't handle big amount of connections. You should decrease number of concurrent connections in Elementum settings (by default it's automatic and is about 800). It's not the number of torrent connections, but another setting to limit overall connections.

Another option is to turn off dht, if you don't need it.

**Comment by elgatito** [comment]
@heartlogic  It will always be the same for you, until you provide any requested information.

**Comment by veniciusgp** [comment]
Dear,

Not if it is already known but when the DHT feature is activated there is intense activity on my network for a few minutes that causes my modem to crash during that time. After a few minutes everything is normalized. If the DHT feature is disabled everything works fine. Is this fact already known? Do I need to send any logs? I am using the latest version (0.71)

**Comment by anacrolix** [comment]
It is known. There's a fix in place in the form of connection tracking. There may be an option exposed in the Elementum UI for it. @elgatito or other regular users may have more info. If you can find what setting works for you that would be ideal.

**Comment by veniciusgp** [comment]
> connection limit is available in Elementum settings, by default it's about 800.

Thanks for listening

With active DHT and UTP I only get stability on my network by limiting the number of connections to a value of 40 or less and this brings me problem with low download speed.

The elementum project is very promising but these problems makes me reuse Quasar. Although discontinued, it uses a much more stable torrent engine that makes it possible to always use the maximum of my internet connection without stopping it.

I hope one day these limitations of the torrent engine will be solved.

Best Regards

**Comment by MichaelAndoid** [comment]
Hi after many months using Quasar because having this problem to in 2 android boxs, know with the new migration to libtorrent in V0.1.20 i can use elementum without ptoblems.
And works ok with the default settings, max speed connection.

Thanks for this migration ;)
